baseband/guppi/payload.py

```diff
@@ -0,0 +1,239 @@
+        else:
+            mode = fh.mode.replace('b', '')
+            offset = fh.tell()
+            words = np.memmap(
+                fh, mode=mode, dtype=cls._dtype_word, offset=offset,
+                shape=(None if payload_nbytes is None
+                       else (payload_nbytes // cls._dtype_word.itemsize,)))
+            fh.seek(offset + words.size * words.dtype.itemsize)
+        return cls(words, header=header, **kwargs)
+
+    @classmethod
+    def fromdata(cls, data, header=None, bps=8, time_ordered=True):
+        """Encode data as a payload.
+
+        Parameters
+        ----------
+        data : `~numpy.ndarray`
+            Data to be encoded. The last dimension is taken as the number of
+            channels.
+        header : `~baseband.guppi.GUPPIHeader`, optional
+            If given, used to infer the ``bps`` and ``time_ordered``.
+        bps : int, optional
+            Bits per elementary sample, used if ``header`` is `None`.
+            Default: 8.
+        time_ordered : bool, optional
+            Whether data are time-ordered, used if ``header`` is `None`.
+            Default: `True`.
+        """
+        if header is not None:
+            bps = header.bps
+            time_ordered = header.time_ordered
+        sample_shape = data.shape[1:]
+        complex_data = data.dtype.kind == 'c'
+        try:
+            encoder = cls._encoders[bps]
+        except KeyError:
+            raise ValueError("{0} cannot encode data with {1} bits"
+                             .format(cls.__name__, bps))
+        # If time-ordered, switch to (nchan, nsample, npol); otherwise use
+        # (nsample, nchan, npol)
+        if time_ordered:
+            data = data.transpose(2, 0, 1)
+        else:
+            data = data.transpose(0, 2, 1)
+        if complex_data:
+            data = data.view((data.real.dtype, (2,)))
+        words = encoder(data).ravel().view(cls._dtype_word)
+        return cls(words, sample_shape=sample_shape, bps=bps,
+                   complex_data=complex_data, time_ordered=time_ordered)
+
+    def __len__(self):
+        """Number of samples in the payload."""
+        return self.nbytes * 8 // self._bpfs // (
+            self.sample_shape.nchan if self.time_ordered else 1)
+
+    def __getitem__(self, item=()):
+        #  GUPPI data may be stored as (nsample, nchan, npol) or, if
+        # "time-ordered", (nchan, nsample, npol), both of which require
+        # reshaping to get the usual order of (nsample, npol, nchan).
+        decoder = self._decoders[self._coder]
+
+        # If we want to decode the entire dataset.
+        if item is () or item == slice(None):
+            data = decoder(self.words)
+            if self.complex_data:
+                data = data.view(self.dtype)
+            if self.time_ordered:
+                # Reshape to (nchan, nsample, npol); transpose to usual order.
+                return (data.reshape(self.sample_shape[1], -1,
```

Comment:
Here and below, can we just use `self.sample_shape.nchan` etc., i.e., use the fact that we have a namedtuple?

Reply:
I originally did that for compactness, but changed.